File: app/src/rendering/__tests__/baselineFixtures.ts
/**
 * Loader for the Layer-1 baseline JSON files that ship under
 * pipeline/tests/snapshots/. They are full Layer-1 emissions for real,
 * public-domain Italian legal texts (Akoma Ntoso XML and IPZS EPUB
 * backends), so they make excellent integration fixtures for Layer 2 — the
 * rendering layer is exercised against the same complexity Layer 1 produces.
 *
 * The committed baselines add a few `_baseline_*` debug fields at the top
 * level (id, source path, structural verdict). Those violate the schema's
 * `additionalProperties: false` rule, so we strip them before validating.
 *
 * The fixtures live OUTSIDE app/ and we read them from disk at test time
 * rather than copying their text into the Layer-2 tree (the text comes from
 * statutes that are public domain, but we still avoid duplication).
 */

import { readFileSync } from 'node:fs';
import path from 'node:path';

import { parseDocument } from '../../consumption';
import type { ScabopdfDocument } from '../../consumption';

const SNAPSHOTS_DIR = path.resolve(
  __dirname,
  '..',
  '..',
  '..',
  '..',
  'pipeline',
  'tests',
  'snapshots',
);

export function loadBaselineDocument(filename: string): ScabopdfDocument {
  const fullPath = path.join(SNAPSHOTS_DIR, filename);
  const raw = readFileSync(fullPath, 'utf8');
  const parsed = JSON.parse(raw) as Record<string, unknown>;

  // Strip top-level _baseline_* debug fields the capture script adds for
  // traceability. They would otherwise fail additionalProperties: false.
  const stripped: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(parsed)) {
    if (!key.startsWith('_baseline_')) {
      stripped[key] = value;
    }
  }
  // The capture script also strips document_id to keep the baselines stable
  // byte-for-byte across runs (it is otherwise a random per-run UUID). The
  // schema requires it; inject a deterministic placeholder so validation
  // passes.
  if (stripped.document_id === undefined) {
    stripped.document_id = '00000000-0000-4000-8000-000000000000';
  }

  const result = parseDocument(stripped);
  if (!result.ok) {
    throw new Error(
      `Baseline ${filename} did not validate: ${result.error.message}`,
    );
  }
  return result.document;
}

/**
 * The fixtures used by the rendering tests. Chosen to cover:
 * - smallest XML AKN + EPUB IPZS fixtures (fast happy path),
 * - mid-size fixtures (more typographic variety),
 * - legge_capitali XML (exercises AMENDMENT / QUOTED_TEXT_* /
 *   UPDATE_BLOCK — categories introduced in schema 0.7.0).
 */
export const BASELINE_FIXTURES = [
  'xml_akn_baseline_legge_56_2007.json',
  'xml_akn_baseline_legge_gelli_bianco.json',
  'xml_akn_baseline_legge_capitali.json',
  'epub_ipzs_baseline_legge_56_2007.json',
  'epub_ipzs_baseline_legge_gelli_bianco.json',
] as const;

File: app/ios/ScaboCore/Tests/ScaboCoreTests/BaselineFixtures.swift
//
//  BaselineFixtures.swift
//  ScaboCoreTests
//
//  Loader for the Layer-1 baseline JSON files committed under
//  pipeline/tests/snapshots/. Faithful translation of
//  `app/src/rendering/__tests__/baselineFixtures.ts`.
//
//  They are full Layer-1 emissions for real, public-domain Italian legal texts
//  (Akoma Ntoso XML and IPZS EPUB backends), so they exercise the rendering layer
//  against the same complexity Layer 1 produces.
//
//  Like the TS, the fixtures live OUTSIDE the app/library tree and are read from
//  disk at test time rather than copied in (avoiding multi-MB duplication). The
//  TS resolves them from `__dirname`; here the snapshots directory is derived from
//  `#filePath` of this source — this file lives at
//  <repo>/app/ios/ScaboCore/Tests/ScaboCoreTests/BaselineFixtures.swift, so six
//  parent steps reach the repo root. This evaluates at build time to the local
//  checkout path, which is correct for `swift test` run in the same sandbox.
//
//  Two TS preprocessing steps and their Swift equivalents:
//   * The capture script adds top-level `_baseline_*` debug fields, which the TS
//     strips because `additionalProperties: false`. Swift's `JSONDecoder` ignores
//     unknown keys by default, so no stripping is needed.
//   * The capture script omits `document_id` (otherwise a random per-run UUID) to
//     keep baselines byte-stable; the schema requires it. As in the TS, a
//     deterministic placeholder is injected before parsing.
//
//  If the snapshots directory is absent (e.g. a partial checkout), the loader
//  throws `XCTSkip` so the suite stays green, per the project fixture convention
//  in CLAUDE.md. A present-but-invalid baseline is a real failure, not a skip.
//

import XCTest
@testable import ScaboCore

enum BaselineFixtureError: Error {
    case parseFailed(String, DocumentLoadError)
}

let SNAPSHOTS_DIR: URL = {
    var url = URL(fileURLWithPath: #filePath)
    for _ in 0..<6 { url.deleteLastPathComponent() }
    return url.appendingPathComponent("pipeline/tests/snapshots")
}()

/// Loads and parses a committed Layer-1 baseline into a `ScabopdfDocument`.
/// Throws `XCTSkip` when the file is missing.
func loadBaselineDocument(_ filename: String) throws -> ScabopdfDocument {
    let url = SNAPSHOTS_DIR.appendingPathComponent(filename)
    guard FileManager.default.fileExists(atPath: url.path) else {
        throw XCTSkip(
            "Baseline \(filename) not present at \(url.path). "
            + "It ships under pipeline/tests/snapshots/ on a full checkout."
        )
    }

    let raw = try Data(contentsOf: url)
    // Inject document_id when absent (the capture script strips it). _baseline_*
    // top-level fields are left in place — the decoder ignores unknown keys.
    var object = try JSONSerialization.jsonObject(with: raw) as! [String: Any]
    if object["document_id"] == nil {
        object["document_id"] = "00000000-0000-4000-8000-000000000000"
    }
    let injected = try JSONSerialization.data(withJSONObject: object)

    switch parseDocument(injected) {
    case .success(let document, _):
        return document
    case .failure(let error):
        throw BaselineFixtureError.parseFailed(filename, error)
    }
}

/// The fixtures used by the rendering layout tests. Chosen to cover the smallest
/// XML AKN + EPUB IPZS fixtures (fast happy path), mid-size fixtures (more
/// typographic variety), and legge_capitali (AMENDMENT / QUOTED_TEXT_* /
/// UPDATE_BLOCK — categories introduced in schema 0.7.0).
let BASELINE_FIXTURES: [String] = [
    "xml_akn_baseline_legge_56_2007.json",
    "xml_akn_baseline_legge_gelli_bianco.json",
    "xml_akn_baseline_legge_capitali.json",
    "epub_ipzs_baseline_legge_56_2007.json",
    "epub_ipzs_baseline_legge_gelli_bianco.json",
]
